### PR TITLE
fix(source): error handling in SourceParser::parse

### PR DESCRIPTION
**Pull request by TennyZhuang**
Signed-off-by: TennyZhuang <zty0826@gmail.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Since we will ignore all parser errors, the original implementation of `SourceParser::parse` may introduce inconsistent `StreamChunk`.

The PR introduced `pop` method on all `ArrayBuilder` implementations, so that we can recover a `DataChunkBuilder` to its consistent state on error.

`pop` is the only way to resolve the problem without introduce unnecessary overhead.

I will add some tests later.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

#5247

### CODE REVIEW

File: src/source/src/parser/mod.rs [hunk 1]
```diff
@@ -88,51 +88,144 @@ pub struct SourceStreamChunkRowWriter<'a> {
...
+struct OpActionUpdate;
+
+impl OpAction for OpActionUpdate {
+    type Output = (Datum, Datum);
+
+    const DEFAULT_OUTPUT: Self::Output = (None, None);
+
+    #[inline(always)]
+    fn apply(builder: &mut ArrayBuilderImpl, output: (Datum, Datum)) -> Result<()> {
+        builder.append_datum(&output.0)?;
```
Comment (TennyZhuang) [inline src/source/src/parser/mod.rs:158]: There is a bug, the `builder` may failed on the intermediate state...

I will use unwrap here

File: src/source/src/parser/mod.rs [hunk 2]
```diff
@@ -88,51 +88,144 @@ pub struct SourceStreamChunkRowWriter<'a> {
...
+
+struct OpActionInsert;
+
+impl OpAction for OpActionInsert {
+    type Output = Datum;
+
+    const DEFAULT_OUTPUT: Self::Output = None;
+
+    #[inline(always)]
+    fn apply(builder: &mut ArrayBuilderImpl, output: Datum) -> Result<()> {
```
Comment (TennyZhuang) [inline src/source/src/parser/mod.rs:111]: I will remove the `Result` and panic on the failure...

File: src/common/src/array/list_array.rs [hunk 3]
```diff
@@ -112,6 +112,19 @@ impl ArrayBuilder for ListArrayBuilder {
...
     }
 
+    fn pop(&mut self) -> Option<()> {
+        if self.bitmap.pop().is_some() {
+            let start = self.offsets.pop().unwrap();
+            let end = *self.offsets.last().unwrap();
+            for _ in end..start {
+                self.value.pop().unwrap()
+            }
+            Some(())
```
Comment (BugenZhao) [inline src/common/src/array/list_array.rs:122]: Also update the `len`?

File: src/source/src/parser/mod.rs [hunk 4]
```diff
@@ -88,51 +88,144 @@ pub struct SourceStreamChunkRowWriter<'a> {
...
-                let datum = if desc.skip_parse { None } else { f(desc)? };
-                builder.append_datum(&datum)?;
+            .enumerate()
+            .try_for_each(|(idx, (desc, builder))| -> Result<()> {
+                let output = if desc.skip_parse {
+                    A::DEFAULT_OUTPUT
+                } else {
+                    f(&desc)?
+                };
+                A::apply(builder, output)?;
```
Comment (BugenZhao) [inline src/source/src/parser/mod.rs:197]: Is it possible that `apply` fails and leaves a dirty state in the array builder? 🤣 Rollbacking cannot help in this case.

**Review by BugenZhao — COMMENTED** [review]
Generally LGTM!

File: src/source/src/parser/mod.rs (hunk 4 above)
Comment (TennyZhuang) [reply src/source/src/parser/mod.rs]: Now apply will not failed.

**Comment by jon-chuang** [comment]
LGTM :)

**Review by tabVersion — APPROVED** [review]
lgtm :-)

**Comment by TennyZhuang** [comment]
@mergify requeue

**Pull request merged**